src/main/java/org/iotdata/constants/dml/TagsQueries.java

```diff
@@ -44,27 +86,43 @@ public class TagsQueries {
 			?resultNode a msr:Measure ;
 				msr:Unit aiotp2:bpm ;
 				msr:hasNumericalValue ?heartRate .
+			
+			FILTER (?heartRate != 0)
 			}
+			GROUP BY ?identifier
 			""";
 
-	public static final String SELECT_DAYS_WITH_WATCH_BATTERY_LEVELS = """
-			SELECT ?timeStamp ?batteryLevel ?identifier
+	public static final String SELECT_ALARM_EVENTS  = """
+			SELECT ?identifier
+					?startTime
+					(?timeStamp AS ?finishTime)
+					?duration
 			WHERE {
-			?subject schema:identifier ?identifier ;
-				sosa:hosts ?sensor .
-				
-			?sensor sosa:madeObservation ?observation .
-			
-			?observation a sosa:Observation ;
-				sosa:resultTime ?timeStamp ;
-				sosa:hasResult ?resultNode .
-				
-			?resultNode a aiotp2:TagMetadataResult ;
-				aiotp2:hasWatchBatteryLevel ?batteryLevel .
+				{
+					SELECT ?timeStamp
+					   	   ?alarmStatus
+					       ?identifier
+					WHERE {
+					?subject schema:identifier ?identifier ;
+						sosa:hosts ?sensor .
+
+					?sensor sosa:madeObservation ?observation .
+
+					?observation a sosa:Observation ;
+						sosa:resultTime ?timeStamp ;
+						sosa:hasResult ?resultNode .
+						
+					?resultNode a aiotp2:TagMetadataResult ;
+						aiotp2:hasAlarm ?alarmStatus .
+					}
+				}
+				BIND(func:mapAlarmEvent(?identifier, ?timeStamp, ?alarmStatus) as ?startTime)
+				BIND((?timeStamp - ?startTime) as ?duration)
+				FILTER(?startTime != ?timeStamp && day(?duration) < 1 && SUBSTR(STR(?duration), 1, 3) != "-PT")
 			}
 			""";
 
-	public static final String SELECT_DAYS_WITH_ALARM = """
+	public static final String SELECT_COUNT_ALARM_TRIGGERS = """
```

Comment:
It is to filter out the records with 'negative' duration, couldn't really figure out anything clearer that would work.

Reply:
But if we have a duration computed - wouldn't, for example, the amount of day/minutes/hours or whatever be also negative? Maybe this would be a more "clear" way to filter these out (especially since you use days in the filter as well).

Reply:
Another thing would be to investigate what the negative duration means - to be completely clear that these are out of order observations and not, e.g. some implementation issue.